lib/Dialect/FIRRTL/Transforms/CreateCompanionAssume.cpp

```diff
@@ -0,0 +1,66 @@
+//===- CreateCompanionAssume.cpp - Create companion assume ----------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the CreateCompanionAssume pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+struct CreateCompanionAssumePass
+    : public CreateCompanionAssumeBase<CreateCompanionAssumePass> {
+  void runOnOperation() override {
+    getOperation().walk([](firrtl::AssertOp assertOp) {
+      OpBuilder builder(assertOp);
+      builder.setInsertionPointAfter(assertOp);
+      auto guards = assertOp->getAttrOfType<ArrayAttr>("guards");
+      Operation *assume;
+      bool isUnrOnlyAssert = false;
+      // Regard the assertion as UNR only if "USE_UNR_ONLY_CONSTRAINTS" is
+      // included in the guards.
+      if (guards) {
+        isUnrOnlyAssert = llvm::any_of(guards, [](Attribute attr) {
+          StringAttr strAttr = dyn_cast<StringAttr>(attr);
+          return strAttr && strAttr.getValue() == "USE_UNR_ONLY_CONSTRAINTS";
+        });
+      }
+
+      // If so, use AssumeEdgedPredicateIntrinsicOp.
+      if (isUnrOnlyAssert)
+        assume = builder.create<firrtl::AssumeEdgedPredicateIntrinsicOp>(
+            assertOp.getLoc(), assertOp.getPredicate(), assertOp.getEnable(),
```

Comment:
Companion assume's -- both kinds -- don't have the message or substitution arguments today, see:

https://github.com/dtzSiFive/firrtl-verif-example/blob/master/assert/conc_true-guard_unr-fmt_%22sva%22-msg_%22message%22-name_%22label%22/output.sv
https://github.com/dtzSiFive/firrtl-verif-example/blob/master/assert/conc_true-guard_no-fmt_%22sva%22-msg_%22message%22-name_%22label%22/output.sv

Does seem nice to have them, just a detail to keep in mind.

Reply:
Yeah, thank you for pointing out! For normal assume I think it's totally Ok to have substitutions. For UNROnly assume, probably they are not enabled in the normal simulation but I agree that the behavior is uncertain.